File: TSRefresh/TSRefresh/Classes/TSRefreshHeader.m
//
//  TSRefreshHeader.m
//  TSRefresh
//
//  Created by 小铭 on 2018/8/21.
//  Copyright © 2018年 caiqr. All rights reserved.
//

#import "TSRefreshHeader.h"

@interface TSRefreshHeader()<TSRefreshHeaderProtocol>

@property (nonatomic, weak) id <TSRefreshHeaderProtocol> delegate;

@end


@implementation TSRefreshHeader

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        
        self.delegate = self;
    }
    return self;
}

- (void)setState:(MJRefreshState)state {
    MJRefreshCheckState
    
    if (state == MJRefreshStateIdle) {
        if (oldState != MJRefreshStateRefreshing) {
            if ([self.delegate respondsToSelector:@selector(tsToNormalState)]) {
                [self.delegate tsToNormalState];
            }
        } else {
            if ([self.delegate respondsToSelector:@selector(tsToRefreshingState)]) {
                [self.delegate tsToRefreshingState];
            }
        }
    } else if (state == MJRefreshStateRefreshing) {
        if ([self.delegate respondsToSelector:@selector(tsToRefreshingState)]) {
            [self.delegate tsToRefreshingState];
        }
    } else if (state == MJRefreshStatePulling) {
        if ([self.delegate respondsToSelector:@selector(tsToPullingState)]) {
            [self.delegate tsToPullingState];
        }
    } else if (state == MJRefreshStateWillRefresh) {
        if ([self.delegate respondsToSelector:@selector(tsToWillRefreshState)]) {
            [self.delegate tsToWillRefreshState];
        }
    }
}

- (void)prepare {
    [super prepare];
    
    if ([self.delegate respondsToSelector:@selector(tsContentHeight)]) {
        self.mj_h = [self.delegate tsContentHeight];
    }
}

- (void)setPullingPercent:(CGFloat)pullingPercent {
    [super setPullingPercent:pullingPercent];
    if ([self.delegate respondsToSelector:@selector(tsChangePullingPercent:)]) {
        [self.delegate tsChangePullingPercent:pullingPercent];
    }
}

@end

File: TSRefresh/TSRefresh/TSRefreshHeader.swift
//
//  TSRefreshHeader.swift
//  TSRefresh
//
//  Created by 小铭 on 2018/6/26.
//  Copyright © 2018年 caiqr. All rights reserved.
//

import UIKit

@objc public protocol TSRefreshHeaderProtocol {
    /// 状态变成.idle
    @objc optional func tsToNormalState()
    /// 状态变成.refreshing
    @objc optional func tsToRefreshingState()
    /// 状态变成.pulling
    @objc optional func tsToPullingState()
    /// 状态变成.willRefresh
    @objc optional func tsToWillRefreshState()
    /// 下拉高度／触发高度 值改变
    @objc optional func tsChangePullingPercent(percent: CGFloat)
    
    /// 控件的高度
    ///
    /// - Returns: 控件的高度
    @objc optional func tsContentHeight() -> CGFloat
}

open class TSRefreshHeader: MJRefreshStateHeader{
    
    public var tsSubProtocol: TSRefreshHeaderProtocol? {
        get { return self as? TSRefreshHeaderProtocol }
    }
    
    override open var state: MJRefreshState {
        didSet {
            if state == .idle {
                if oldValue == .refreshing {
                    
                } else {
                    self.tsSubProtocol?.tsToNormalState?()
                }
            } else if state == .refreshing {
                self.tsSubProtocol?.tsToRefreshingState?()
            } else if state == .pulling {
                self.tsSubProtocol?.tsToPullingState?()
            } else if state == .willRefresh {
                self.tsSubProtocol?.tsToWillRefreshState?()
            }
        }
    }
    
    override open var pullingPercent: CGFloat {
        didSet {
            self.tsSubProtocol?.tsChangePullingPercent?(percent: pullingPercent)
        }
    }
    
    override public init(frame: CGRect) {
        super.init(frame: frame)
    }
    
    open override func prepare() {
        super.prepare()
        if let contentHeight = self.tsSubProtocol?.tsContentHeight?() {
            self.mj_h = contentHeight
        }
    }
    
    override open func layoutSubviews() {
        super.layoutSubviews()
    }
    
    required public init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
}
